### ISSUE TITLE
[Feature]: Adds resources for challenges

### ISSUE THREAD

**Issue by bautistaaa** [typehero/typehero#1021]
### Description

Pointing users to relevant docs  to solve the beginner challenges

**Comment by Mahmoudgalalz** [comment]
how do you think it should be

I have some resources, but how it will work 

I think we should added the resources on each problem, to be more specific
it might be a video of matt explaining Generics or the Docs or any type of resource

is this functionality available?

**Comment by Dun-sin** [comment]
> how do you think it should be
> 
> I have some resources, but how it will work
> 
> I think we should added the resources on each problem, to be more specific it might be a video of matt explaining Generics or the Docs or any type of resource
> 
> is this functionality available?

what about some form of AI, that detects the topic but the better way is when a challenge is being set, a compulsory section should be dropping links to resources, another should be the community dropping resources and they are vetted to be okay and challengers can pick a resource from an existing list.

**Comment by bautistaaa** [comment]
what goes in resources is determined by the challenge author as it will be embedded with the prompt. can probably look at leetcode for inspiration but it’s nothing that needs AI

**Comment by Dun-sin** [comment]
> what goes in resources is determined by the challenge author as it will be embedded with the prompt. can probably look at leetcode for inspiration but it’s nothing that needs AI

yes that's why i said the challenger option is a better option

**Comment by bautistaaa** [comment]
can probably just have an additional drawer next to comments that’s resources

**Comment by Dun-sin** [comment]
> can probably just have an additional drawer next to comments that’s resources

my point isn't the display but how to get it

**Comment by bautistaaa** [comment]
i’m not even talking to you?

**Comment by Dun-sin** [comment]
> i’m not even talking to you?

you don't have to  be rude, i expect you to be civil, there can be misunderstandings and that's why we have conversations, you noticed how i tag your message to show you i'm talking to you?

You didn't do that, tag people next time and talk like you are talking to people and not your slave or someone you are inferior to

**Comment by bautistaaa** [comment]
\

> > i’m not even talking to you?
> 
> you don't have to be rude, i expect you to be civil, there can be misunderstandings and that's why we have conversations, you noticed how i tag your message to show you i'm talking to you?
> 
> You didn't do that, tag people next time and talk like you are talking to people and not your slave or someone that's inferior to you

sorry im at lunch with my kids and gave a short answer. 

in the `challenges` dir is where an author creates a challenge and they will probably end up creating a `resources` file which will be ingested into a new `resources` field on  challenges

**Comment by bautistaaa** [comment]
> > how do you think it should be
> > I have some resources, but how it will work
> > I think we should added the resources on each problem, to be more specific it might be a video of matt explaining Generics or the Docs or any type of resource
> > is this functionality available?
> 
> what about some form of AI, that detects the topic but the better way is when a challenge is being set, a compulsory section should be dropping links to resources, another should be the community dropping resources and they are vetted to be okay and challengers can pick a resource from an existing list.

i dont like the idea of community vetted resources atm as that requires more moderation and we simply dont have anyone to do it. we do have the ability to pin solutions in the solutions tab which is either ideally a pinned solution by the author or one we deem as well thought out but its all adhoc

**Comment by Dun-sin** [comment]
> sorry im at lunch with my kids and gave a short answer.
> 
> in the `challenges` dir is where an author creates a challenge and they will probably end up creating a `resources` file which will be ingested into a new `resources` field on challenges

yes, what i said but more techincal, i haven't had the chance to look at the code yet, but gave my general thought

**Comment by bautistaaa** [comment]
> how do you think it should be
> 
> I have some resources, but how it will work
> 
> I think we should added the resources on each problem, to be more specific it might be a video of matt explaining Generics or the Docs or any type of resource
> 
> is this functionality available?

resources are going to be populated by the author and there is a way to pin solutions that we consider "official"

**Comment by Dun-sin** [comment]
> > > how do you think it should be
> > > I have some resources, but how it will work
> > > I think we should added the resources on each problem, to be more specific it might be a video of matt explaining Generics or the Docs or any type of resource
> > > is this functionality available?
> > 
> > 
> > what about some form of AI, that detects the topic but the better way is when a challenge is being set, a compulsory section should be dropping links to resources, another should be the community dropping resources and they are vetted to be okay and challengers can pick a resource from an existing list.
> 
> i dont like the idea of community vetted resources atm as that requires more moderation and we simply dont have anyone to do it. we do have the ability to pin solutions in the solutions tab which is either ideally a pinned solution by the author or one we deem as well thought out but its all adhoc

understandable, that could be an option when this project has more contributors, looking forward to be a good one in the future.

We can just leave it for now

**Comment by Dun-sin** [comment]
> > how do you think it should be
> > I have some resources, but how it will work
> > I think we should added the resources on each problem, to be more specific it might be a video of matt explaining Generics or the Docs or any type of resource
> > is this functionality available?
> 
> resources are going to be populated by the author and there is a way to pin solutions that we consider "official"

sounds like a good option, how do you consider a solution(do you mean challenge, not sure) official?

**Comment by bautistaaa** [comment]
> > > how do you think it should be
> > > I have some resources, but how it will work
> > > I think we should added the resources on each problem, to be more specific it might be a video of matt explaining Generics or the Docs or any type of resource
> > > is this functionality available?
> > 
> > 
> > resources are going to be populated by the author and there is a way to pin solutions that we consider "official"
> 
> sounds like a good option, how do you consider a solution(do you mean challenge, not sure) official?

not sure if you've seen the site but its much easier if you have access. if you dont you can join the discord and i can provide you an early access token. 

users can share solutions once they have successfully provided one that others can see. admins can pin if they think its appropriate(see screenshot for example)

<img width="592" alt="CleanShot 2023-11-05 at 12 23 08@2x" src="https://github.com/typehero/typehero/assets/3660667/ab8ec02b-08cd-4218-b068-60c7aa7083a7">

**Comment by Dun-sin** [comment]
> > > > how do you think it should be
> > > > I have some resources, but how it will work
> > > > I think we should added the resources on each problem, to be more specific it might be a video of matt explaining Generics or the Docs or any type of resource
> > > > is this functionality available?
> > > 
> > > 
> > > resources are going to be populated by the author and there is a way to pin solutions that we consider "official"
> > 
> > 
> > sounds like a good option, how do you consider a solution(do you mean challenge, not sure) official?
> 
> not sure if you've seen the site but its much easier to envision if you have access. if you dont you can join the discord and i can provide you an early access token.
> 
> users can share solutions once they have successfully provided one that others can see. admins can pin if they think its appropriate(see screenshot for example)
> 
> <img alt="CleanShot 2023-11-05 at 12 23 08@2x" width="592" src="https://user-images.githubusercontent.com/3660667/280552655-ab8ec02b-08cd-4218-b068-60c7aa7083a7.png">

i don't have access yet, joined the waiting list just today, i think i commented on your linkedin post about, just here to know more about the project before getting the access, will join the discord.

That seems reasonable, is it mostly based on stars or is it code review kind of way(doubt this because you said moderation is a problem, but curious either way)

lastly, do you think it's better to continue this conversion on your discord? i think this is kind of causing an unnecessary read for other contributors, whatever is comfortable for you or what you think.

You can tag me with your answer on this, will join real quick

**Comment by Mahmoudgalalz** [comment]
After taking a look at the `challenges` dir, it seems like we can add these resources inside the metadata

and also try to force it, like by creating a pipeline that triggers PR starts with challenges:

to see if there's a resource or not

About the displaying, I think attach it to the drawer, "Learn"

This is all my thoughts about it

**Comment by aadamw** [comment]
I think we can tackle that feature incrementally in 2 steps:

1. Adding an example on the description step. This would indicate that a person creating a new challenge can add additional resources. I think this would be the fastest and easiest solution. No additional code changes would be required except the example:
<img width="1679" alt="Screenshot 2023-11-07 at 12 13 29" src="https://github.com/typehero/typehero/assets/131662061/6d4e9b16-3c3b-42fa-adf4-9762e6fdc216">

2. Adding another step inside the challenge wizard just for adding resources, or appending this form to an already existing one (something like we have for links in profile settings). This would require us to introduce a new table for `ChallengeResources` in db, add a new form to the challenge wizard, display it in the admin panel, and render it in challenge view.

**Comment by bautistaaa** [comment]
> I think we can tackle that feature incrementally in 2 steps or just go with 1 or 2 instantly:
> 
> 1. Adding an example on the description step. This would indicate that a person creating a new challenge can add additional resources. I think this would be the fastest and easiest solution. No additional code changes would be required except the example:
> 
> <img alt="Screenshot 2023-11-07 at 12 13 29" width="1679" src="https://user-images.githubusercontent.com/131662061/281018877-6d4e9b16-3c3b-42fa-adf4-9762e6fdc216.png">
> 2. Adding another step inside the challenge wizard just for adding resources, or appending this form to an already existing one (something like we have for links in profile settings). This would require us to introduce a new table for `ChallengeResources` in db, add a new form to the challenge wizard, display it in the admin panel, and render it in challenge view.

i’m actually axing this feature(wizard) and submission will be going through github. users will submit by following the guidelines in the challenges directory. i believe @dimitropoulos will be adjusting this to account for the extra field. how we want to display is still up in the air but markdown is probably the easiest. how we store it could very well be what you suggest!

**Comment by Mahmoudgalalz** [comment]
@bautistaaa so the wizard is not longer supported, and all challenges will be added through a PR right?
